### PR TITLE
TL-10453: Triplelift adapter for openrtb s2s

### PR DESCRIPTION
**Pull request by Kevin-P-Kerr**
This p.r. defines the adapter which calls the exchange.  In particular, I defined ext objects on request.imp and response.imp.  On the request, we send the `floor` and `inventoryCode`, just as in prebid c2s.  On the response, we have to send the format, so that the prebid s2s server can process the response. 

request to prebid
```
{ "id": "some-request-id",
   "site": {
    "page": "https://www.123greetings.com/events/fathers_day/happy_fathers_day",
    "publisher": {
      "id": "pub-4627517680249670",
      "ext": {
        "country": "US"
      }
    } 
    },
    "imp": [
      {
        "tagid" : "makersalley_plain",
        "id": "some-impression-id",
        "banner": {
          "format": [
            {
              "w": 300,
              "h": 250
            },
            {
              "w": 300,
              "h": 600
            }
          ]
        },
        "ext": {
          "triplelift" : { "inventoryCode":"makersalley_main"} 
        }
      }
    ],
    "tmax": 500
  }
```
request from prebid to tl
```
"{
  "id": "some-request-id",
  "imp": [
    {
      "id": "some-impression-id",
      "banner": {
        "format": [
          {
            "w": 300,
            "h": 250
          },
          {
            "w": 300,
            "h": 600
          }
        ]
      },
      "tagid": "makersalley_plain",
      "ext": {
        "bidder": {
          "inventoryCode": "makersalley_main"
        }
      }
    }
  ],
  "site": {
    "page": "https://www.123greetings.com/events/fathers_day/happy_fathers_day",
    "publisher": {
      "id": "pub-4627517680249670",
      "ext": {
        "country": "US"
      }
    },
    "ext": {
      "amp": 0
    }
  },
  "device": {
    "ua": "curl/7.43.0",
    "ip": "::1"
  },
  "at": 1,
  "tmax": 500
}"
```
response from prebid
```
"{
  "id": "some-request-id",
  "seatbid": [
    {
      "bid": [
        {
          "id": "21524535241656531860",
          "impid": "some-impression-id",
          "price": 3.338,
          "adm": "<script>document.createElement('IMG').src=\"https://tlx.3lift.net/s2s/notify?px=1&ts=1560870964&aid=21524535241656531860&ec=260_9_2763332&n=GqsBaHR0cDovL2V4Y2hhbmdlLXRlc3Rlci50cmlwbGVsaWZ0Lm5ldDo4MDc3L25vdGlmeT9pZD0yMTUyNDUzNTI0MTY1NjUzMTg2MCZwcmljZT0ke0FVQ1RJT05fUFJJQ0V9JmFiaWQ9MTU2MzA5ODUzODIwODY3NjYwNyZhaW1waWQ9MTU2MzA5ODUzODIwODY3NjYwNy0xLW51bGwmdHlwZT1CaWQyMlNlYXRz8gKeAQgAEhQyMTUyNDUzNTI0MTY1NjUzMTg2MBgAIAEohAIwCUABSABQAGABaABwmKMFkAEAmAEAqAH%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F8BsAGPKLgBAMABihrIAY8o4AGKGvABAPgB0imAAogbiAKKGpECAAAAAAAA8D%2BZAmZmZmZmZtY%2FoQIAAAAAAAAAAKgCALACBMgCAtgCAPECZmZmZmZm5j%2F4AoQCkAMA%2BAITiAMAkgMEYmNkZQ%3D%3D\";window.tl_auction_response_644731={\"settings\":{\"advertiser_name\":\"TEST ADVERTISER NAME\",\"viewability\":{},\"impression_pixels\":[\"https:\\/\\/ad.doubleclick22222.net\\/ad\\/N4406.282142.ADARAMEDIAINC.\\/B8015471.114430570;sz=1x1;ord=142248259412135895294\"],\"type\":\"image\",\"additional_data\":{\"uid\":\"8947252401186410784\",\"brid\":\"86424\",\"bmid\":\"260\",\"aid\":\"21524535241656531860\",\"ts\":\"1560870964\",\"sid\":\"9\"},\"logo_image_url\":\"http:\\/\\/ep.yimg.com\\/ay\\/entirelypets\\/animal-planet-bat-dog-costume-large-21.jpg\",\"format_id\":1,\"render_options_bm\":0},\"assets\":[{\"image_id\":-1,\"image_url\":\"https:\\/\\/fublag.files.wordpress.com\\/2009\\/04\\/gen66337.png\",\"heading\":\"Renaissance Curacao Seat 3\",\"caption\":\"Discover the hidden gem of the Caribbean\",\"clickthrough_url\":\"http:\\/\\/ag.yieldoptimizer22222.com\\/ag\\/event?type=click&gi=vHH2MVL_SjnGKP62Cib00jK0s2o-vYF4fY9KN-ZMFBizsiWz6ErQ6BZVg_BuhEdSMwuOZFakknl2X_IMznDSZwult3EAcf-2t5bLa92B_a55wqXvj8LVNcUw5THdrYVDkzGglcgDIGNJBHMXOMtXE8627ymU1e3DANpWp1TQam6TMsQptW1GkKjvN34QhFK0&ref_url=123greetings.com\"}]};</script><script src=\"//ib.3lift.net/ttj?inv_code=makersalley_main\" data-auction-response-id=\"644731\"></script>",
          "adomain": [
            "marriott.com"
          ],
          "iurl": "https://fublag.files.wordpress.com/2009/04/gen66337.png",
          "crid": "260_9_2763332",
          "w": 300,
          "h": 250,
          "ext": {
            "prebid": {
              "type": "native"
            },
            "bidder": {
              "triplelift_pb": {
                "format": 1
              }
            }
          }
        }
      ],
      "seat": "triplelift"
    }
  ],
  "ext": {
    "responsetimemillis": {
      "triplelift": 121
    },
    "tmaxrequest": 500
  }
}"
```

### CODE REVIEW

File: adapters/triplelift/triplelift.go [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+type TripleliftRespExtTriplelift struct {
+	format int `json:"format"`
+}
+
+type TripleliftRespExt struct {
+	Triplelift TripleliftRespExtTriplelift `json:"triplelift_pb"`
+}
+
+func getBidType(ext TripleliftRespExt) (openrtb_ext.BidType, error) {
```
Comment (yuzawa-san) [inline adapters/triplelift/triplelift.go:27]: everything is banner unless it is hdx instream

File: adapters/triplelift/triplelift.go [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        imp.BidFloor = floor 
+    }
+    // no error
+    return nil
+}
+
+func (a *TripleliftAdapter) MakeRequests(request *openrtb.BidRequest) ([]*adapters.RequestData, []error) {
+	errs := make([]error, 0, len(request.Imp))
+	reqs := make([]*adapters.RequestData, 0, 1)
+    // copy the request, because we are going to mutate it
```
Comment (yuzawa-san) [inline adapters/triplelift/triplelift.go]: consistent spacing

File: static/bidder-info/triplelift.yaml [hunk 3]
```diff
@@ -0,0 +1,13 @@
+maintainer:
+  email: "prebid@triplelift.com"
+capabilities:
+  app:
+    mediaTypes:
+      - banner
+      - native
```
Comment (yuzawa-san) [inline static/bidder-info/triplelift.yaml]: remove native here and below

File: adapters/triplelift/triplelift.go (hunk 2 above)
Comment (Kevin-P-Kerr) [reply adapters/triplelift/triplelift.go]: `Running: gofmt -s -w adapters/triplelift/triplelift.go`

**Pull request merged**